### PR TITLE
feat: add pct, min, max rt statistics

### PR DESCRIPTION
**Pull request by maojunxyz**
add (25, 50, 75, 90, 95, 99) pct, min & max statistics

### CODE REVIEW

**Comment by mashuangwei** [comment]
it is very good

**Comment by junoyoon** [comment]
Good PR. 
But I think it should be active only when users want.

image

I think this can be set by providing the checkbox in the perftest configuration (this is not the general case so it should be in detailed setting area in below)

**Comment by donggyu04** [comment]
@maojunxyz

You should consider h2 and mysql both. and please add db.changelog_schema.xml to [here](https://github.com/naver/ngrinder/tree/81d57b136eac646a400aa0a90149e29199885f51/ngrinder-controller/src/main/resources/ngrinder_datachange_logfile) for backward compatibility.

File: ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_H2.xml [hunk 1]
```diff
@@ -73,6 +73,15 @@
 		        last_progress_message varchar(2048),
 		        mean_test_time double,
 		        peak_tps double,
+				errorRate double,
```
Comment (donggyu04) [inline ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_H2.xml:76]: This code is not backwards compatible. add new `db.changelog_schema_xx.xml` [here](https://github.com/naver/ngrinder/tree/81d57b136eac646a400aa0a90149e29199885f51/ngrinder-controller/src/main/resources/ngrinder_datachange_logfile)

Comment (maojunxyz) [reply ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_H2.xml:76]: Got it,  I will fix it.

File: ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue [hunk 2]
```diff
@@ -61,6 +61,60 @@
                             <th v-text="i18n('perfTest.report.errors')"></th>
                             <td>{{ test.errors | numFormat }}</td>
                         </tr>
+                        <tr>
+                            <th v-text="i18n('perfTest.report.errorRate')"></th>
+                            <td>{{ test.errors /(test.tests + test.errors)*100 |  numFormat('0,0.00') }}
```
Comment (imbyungjun) [inline ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue:66]: You better clean up white spaces here like
```vue
<td>{{ test.errors / (test.tests + test.errors) * 100 | numFormat('0,0.00') }}
```

File: ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue [hunk 3]
```diff
@@ -61,6 +61,60 @@
...
+                        </tr>
+                        <tr>
+                            <th v-text="i18n('perfTest.report.minRT')"></th>
+                            <td>{{ test.minRT | numFormat  | numFormat('0,0.00') }}
+                                <code class="ml-1">ms</code>
+                            </td>
+                        </tr>
+                        <tr>
+                            <th v-text="i18n('perfTest.report.maxRT')"></th>
+                            <td>{{ test.maxRT | numFormat  | numFormat('0,0.00') }}
```
Comment (imbyungjun) [inline ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue:114]: Why did you round to decimal by using `numFormat` filter twice?
image

The result what you expected is like capture below?
image

If so, you have to change `numFormat` filter like this.
```vue
<td>{{ test.maxRT | numFormat('0,0.00') }}
```

File: ngrinder-core/src/main/java/net/grinder/SingleConsole.java [hunk 4]
```diff
@@ -857,6 +866,42 @@ protected void updateStatistics(StatisticsSet intervalStatistics, StatisticsSet
...
+			LOGGER.debug("getAdditionalStats() responseTimeList is {}", responseTimeList);
+		}
+
+		int i = 0;
+		double[] rtArray = new double[responseTimeList.size()];
+		for (double responseTime : responseTimeList) {
+			rtArray[i++] = responseTime;
+		}
+
+		Arrays.sort(rtArray);
```
[the remark this answers is not in the document]
Comment (maojunxyz) [reply ngrinder-core/src/main/java/net/grinder/SingleConsole.java:887]: LGTM, thanks.

File: ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue (hunk 2 above)
Comment (maojunxyz) [reply ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue:66]: LGTM, thanks. I will fix it.

File: ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue (hunk 3 above)
Comment (maojunxyz) [reply ngrinder-frontend/src/js/components/perftest/report/DetailReport.vue:114]: tks, I will fix it.

File: ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_34.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+            <column name="pct50RT" type="double"/>
+            <column name="pct75RT" type="double"/>
+            <column name="pct90RT" type="double"/>
+            <column name="pct95RT" type="double"/>
+            <column name="pct99RT" type="double"/>
+            <column name="minRT" type="double"/>
+            <column name="maxRT" type="double"/>
+        </addColumn>
+    </changeSet>
+</databaseChangeLog>
```
[the remark this answers is not in the document]
Comment (maojunxyz) [reply ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_34.xml:18]: thanks,I will fix it.

File: ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_34.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd">
+    <changeSet author="ngrinder.3.5.6" id="34" dbms="h2, mysql">
```
Comment (donggyu04) [inline ngrinder-controller/src/main/resources/ngrinder_datachange_logfile/db.changelog_schema_34.xml:5]: please fix author to ngrinder.3.5.8

File: ngrinder-controller/src/main/resources/messages_en.properties [hunk 7]
```diff
@@ -228,6 +228,15 @@ perfTest.report.meantime=Mean Test Time
 perfTest.report.totalTests=Executed Tests
 perfTest.report.successfulTests=Successful Tests
 perfTest.report.errors=Errors
+perfTest.report.errorRate=Error Rate
```
Comment (donggyu04) [inline ngrinder-controller/src/main/resources/messages_en.properties:231]: @imbyungjun 

We should add `messages_kr.properties` before merging.

**Comment by donggyu04** [comment]
@maojunxyz 

are you reflecting [above comment](https://github.com/naver/ngrinder/pull/917#issuecomment-1286396866)?

**Comment by maojunxyz** [comment]
> Good PR. But I think it should be active only when users want.
> 
> image
> 
> I think this can be set by providing the checkbox in the perftest configuration (this is not the general case so it should be in detailed setting area in below)

But I think this is a very common indicator, like jmeter's aggregate report, the collection dashboard of grafana+influxdb provides it by default.

**Comment by maojunxyz** [comment]
> @maojunxyz
> 
> are you reflecting [above comment](https://github.com/naver/ngrinder/pull/917#issuecomment-1286396866)?

I want to keep it simple and don't want the user to have too many options making it harder to use.

**Comment by maojunxyz** [comment]
> @maojunxyz
> 
> are you reflecting [above comment](https://github.com/naver/ngrinder/pull/917#issuecomment-1286396866)?

I want to keep it simple and don't want the user to have too many options making it harder to use.

**Comment by junoyoon** [comment]
@maojunxyz 
Sorry but you're not the maintainer.
You can want your way but you should follow our guideline if you want merge this into main stream.

I think your requirement is not so necessary for many of users but it gives another calculation burden to ngrinder even for the users who don't need this. We want to avoid this.

I'll find out the better guidance to resolve this.

**Pull request closed without merging**